### ISSUE TITLE
Create a "First To Solve" report in Ceremonies-requested format

### ISSUE THREAD

**Issue by clevengr** [pc2ccs/pc2v9#791]
**Feature Description**: 

Add a new Report that generates "First To Solve" information in the format requested every year by the WF Ceremonies Committee.

Every year the Committee asks for FTS information in a specific format:  one line for each solved problem showing the time of earliest solution for that problem and the team who solved it, _ordered by increasing time of first solution_.    This report data is used both by the Ceremonies Committee and the Judges who announce "First To Solve" awards during the Closing Ceremony.

The report should also include, for each FTS team, a list of the problems which that team was First-to-Solve.

For example:
```
FIRST TO SOLVE IN SOLUTION ORDER
------------------------------------------
PROBLEM            TEAM         TIME
    C              UCLA           7
    F              Texas         10
    A              MIT           13
    G              UCLA          17
    B              Toronto       23
    D              <not solved>
    E              <not solved>

FIRST TO SOLVE BY TEAMS
-----------------------------
TEAM       FIRST TO SOLVE
MIT          A
Texas        F
Toronto      B
UCLA         C, G
Not Solved:  D, E
```
Report columns should be properly tabulated (vertically aligned).

**Have you considered other ways to accomplish the same thing?** 
Yes.  There are existing reports that effectively contain all this information.  However, the information is not organized in the above format; it requires the reader to cross-reference back and forth to figure out the two basic things the Committee wants to know:  who solved each problem first, in what order; and, what problems did each team solve first?   It requires effort to extract this information from existing reports, whereas a new report organized as above would provide the information directly. 

**Do you have any specific suggestions for how your feature would be ***implemented*** in PC^2?** 
- Use the existing Report infrastructure (template) -- just create a new report organized as above.
- Add a new item to the Reports drop-down menu:  **First To Solve**

**Additional context**: 
While currently this report is only needed at World Finals, and PC2 is not (yet) the Primary CCS there,  DOMjudge already produces such a report and PC2 should keep up in anticipation of perhaps being "promoted" to Primary.  Also, the same people who want this report at WF are also increasingly involved in other contests (e.g. NAC) where it would be useful.

**Comment by lane55** [comment]
This Issue should/will be based on the Awards report infrastructure #383
Work can start on this once #383 is merged
Est LOE is one hour to code and unit test

**Comment by clevengr** [comment]
> This Issue should/will be based on the Awards report infrastructure https://github.com/pc2ccs/pc2v9/issues/383

I guess I don't understand why [Issue 383 ](https://github.com/pc2ccs/pc2v9/issues/383) is particularly relevant to this Issue.  There's no need as far as I can see for things like Messages and MessageListeners (which are integral to 383).  Is there some reason for not using the exact same structure as what is used in `core/report/FastestSolvedReport.java` and `core/report/FastestSolvedSummaryReport.java` -- that is, create a new class `core/report/FirstToSolveReport.java` which does the same thing as those existing reports, which is basically
- call `contest.getRuns()` and `contest.getProblems()`
- search through the runs and problems looking for the ones that need to be put into the new report?

It seems to me that 80-90% of what's needed is already being done in those classes; I don't quite see why a new report couldn't/shouldn't be done the same way...

**Comment by johnbrvc** [comment]
> 

Actually, issue #383 does seem relevant.  Maybe you're confusing this with issue #758 (PR #759) which has already been approved?

**Comment by clevengr** [comment]
> > 
> 
> Actually, issue #383 does seem relevant. Maybe you're confusing this with issue #758 (PR #759) which has already been approved?

No, I'm not confusing it (maybe Doug is, in his original comment?)   My point was that this issue is to create a new report which contains much of the same information as that which is already contained in the existing reports generated by `core/report/FastestSolvedReport` and `core/report/FastestSolvedSummaryReport` -- with the information in the new report simply being formatted in the way that the Ceremonies committee would like to have it presented.  I think that the simplest way to do that is to use the same structure as what is used in those two classes and simply reformat the output as requested.  I don't see why the added complexity of the structure created by Issue #383 is needed, especially since that Issue is about creating a JSON file (and in addition it uses a new notification mechanism which isn't really needed for this Issue).

**Comment by clevengr** [comment]
Ok, upon further review.... I withdraw my comments about MessageRecords and MessageListeners having anything to do with this issue; I was clearly thinking about the wrong PR.

I do NOT, however, withdraw my assertion that the following statement is inaccurate and misleading:
```
This Issue should/will be based on the Awards report infrastructure https://github.com/pc2ccs/pc2v9/issues/383
Work can start on this once https://github.com/pc2ccs/pc2v9/issues/383 is merged
```
I don't see anything in Issue #383 that is required for this current issue.  Issue #383 (and its corresponding PR #578) have to do with adding support for creation of a CLICS-compatible JSON Awards report.  That PR adds new classes `CLICSAward`, `CLICSAwardReport`, and `AwardJSON`, along with new utilities and support classes to support generation of a `CLICSAwardReport`.   While a `CLICSAwardReport` certainly contains most (but not all!)  the information requested by the Ceremonies committee as represented in THIS Issue, so (as I pointed out above) do classes `FastestSolvedReport` and `FastestSolvedSummaryReport`.  However, just like those classes, `CLICSAwardReport` doesn't contain the information _in the requested format_.   Further, the `CLICSAwardReport` focusses on creating a JSON format, which is unnecessary (and unwanted) for the feature request which formed the foundation of THIS issue.

My assertion is that trying to use the changes represented by Issue #383 / PR #578 is at best overkill and is not a good solution.  As I indicated, I think a very reasonable solution is to clone an existing report (such as `FastestSolvedSummaryReport`, which has access to all the required data), and simply reformat the data in the requested way.  No need to involve anything related to `CLICSAwardReport` or its accompanying classes.

Ultimately of course, whoever gets around to actually coding up an implementation (PR) for this issue will likely determine what approach is taken... assuming they can get two other people to agree.   The Issue itself is of course low priority so I don't really see the likelihood of anyone working on it anytime soon...

**Comment by lane55** [comment]
Why #383 ?   Well the First to Solve is required to create the awards.json, so using that code
to be the basis of the report is the proper design.   There should be only one place to assign/generate
first to solve and that is in the #383.     Once #383 is done then that can be the input to generate
the first to solve report.
